File: mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperOffices.java
package com.mifos.api.local.databasehelper;

import android.os.AsyncTask;

import com.mifos.objects.organisation.Office;
import com.mifos.objects.organisation.OfficeOpeningDate;
import com.raizlabs.android.dbflow.sql.language.SQLite;

import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.Subscriber;

/**
 * Created by Rajan Maurya on 7/7/16.
 */
@Singleton
public class DatabaseHelperOffices {


    @Inject
    public DatabaseHelperOffices() {

    }


    public Observable<Void> saveAllOffices(final List<Office> offices) {
        AsyncTask.THREAD_POOL_EXECUTOR.execute(new Runnable() {
            @Override
            public void run() {

                for (Office office : offices) {

                    OfficeOpeningDate officeOpeningDate = new OfficeOpeningDate();
                    officeOpeningDate.setOfficeId(office.getId());
                    officeOpeningDate.setYear(office.getOpeningDate().get(0));
                    officeOpeningDate.setMonth(office.getOpeningDate().get(1));
                    officeOpeningDate.setDay(office.getOpeningDate().get(2));

                    office.setOfficeOpeningDate(officeOpeningDate);
                    office.save();
                }

            }
        });
        return null;
    }


    public Observable<List<Office>> readAllOffices() {
        return Observable.create(new Observable.OnSubscribe<List<Office>>() {
            @Override
            public void call(Subscriber<? super List<Office>> subscriber) {

                List<Office> offices = SQLite.select()
                        .from(Office.class)
                        .queryList();

                for (int i = 0; i < offices.size(); i++) {
                    offices.get(i).setOpeningDate(Arrays.asList(
                            offices.get(i).getOfficeOpeningDate().getYear(),
                            offices.get(i).getOfficeOpeningDate().getMonth(),
                            offices.get(i).getOfficeOpeningDate().getDay()));
                }

                subscriber.onNext(offices);
            }
        });
    }
}

File: core/datastore/src/main/java/com/mifos/core/databasehelper/DatabaseHelperOffices.kt
package com.mifos.core.databasehelper

import android.os.AsyncTask
import com.mifos.core.objects.organisation.Office
import com.mifos.core.objects.organisation.OfficeOpeningDate
import com.raizlabs.android.dbflow.sql.language.SQLite
import rx.Observable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Rajan Maurya on 7/7/16.
 */
@Singleton
class DatabaseHelperOffices @Inject constructor() {
    fun saveAllOffices(offices: List<Office>): Observable<Void>? {
        AsyncTask.THREAD_POOL_EXECUTOR.execute {
            for (office in offices) {
                val officeOpeningDate = OfficeOpeningDate()
                officeOpeningDate.officeId = office.id
                officeOpeningDate.year = office.openingDate[0]
                officeOpeningDate.month = office.openingDate[1]
                officeOpeningDate.day = office.openingDate[2]
                office.officeOpeningDate = officeOpeningDate
                office.save()
            }
        }
        return null
    }

    fun readAllOffices(): Observable<List<Office>> {
        return Observable.create<List<Office>> { subscriber ->
            val offices = SQLite.select()
                .from(Office::class.java)
                .queryList()
            for (i in offices.indices) {
                offices[i].openingDate = listOf(
                    offices[i].officeOpeningDate?.year,
                    offices[i].officeOpeningDate?.month,
                    offices[i].officeOpeningDate?.day
                )
            }
            subscriber.onNext(offices)
        }
    }
}